src/psyclone/psyad/tl2ad.py

```diff
@@ -262,29 +262,56 @@ def generate_adjoint(tl_psyir, active_variables):
-            visibility=kernel_sym.visibility)
-        container.symbol_table.remove(kernel_sym)
-        routine.name = adj_kernel_sym.name
-    else:
-        routine.name = routine.symbol_table.next_available_name(
-            _create_adjoint_name(routine.name))
+    # TODO issue #1820 support separate LFRic-specific and generic
+    # implementations as they would have different solutions here.
+
+    # TODO issue #1820 if this is an LFRic-specific implementation and
+    # the metadata code points to an interface then adjoint all
+    # routines specified in the interface. If it points to a routine
+    # then only translate that routine. If this is a generic
+    # implementation then only support one routine as we don't know
+    # which one to support.
+
+    # TODO issue #1800 if there are multiple
+    # modules/subroutines/program implementations in the file then
+    # raise an exception unless a particular name is specified (by
+    # e.g. command line -kernel_name=xyz. The name can be for a routine
+    # or an interface.)
+
+    # Until we know whether this is meant to be a generic or
+    # LFRic-specific kernel (issue #1820) and, for LFRic, can specify
+    # the particular kernel metadata if multiple versions exist and
+    # then read kernel metadata to determine whether it points to a
+    # kernel or interface (issue #1807), we simply assume that we
+    # should allow multiple routines as they imply an interface. We
+    # further assume that the implementation of the routines in the
+    # interface use the same variable names which allows us to
+    # continue to use a single command line list of active
+    # variables. This is the case for the implementations we care
+    # about but in general may not be the case. Issue #1595 should
+    # help fix this problem as it would only be arguments that would
+    # need to have the same names.
+
+    for routine in routines:
+
+        # We need to re-name the kernel routines. We have to take care
+        # in case we've been supplied with a bare program/subroutine
+        # rather than one or more subroutines within a module (which
+        # we will get with LFRic mixed precision kernels).
+        if container:
+            kernel_sym = container.symbol_table.lookup(routine.name)
+            adj_kernel_name = _create_adjoint_name(routine.name)
+            # A symbol's name is immutable so create a new RoutineSymbol
+            adj_kernel_sym = container.symbol_table.new_symbol(
+                adj_kernel_name, symbol_type=RoutineSymbol,
+                visibility=kernel_sym.visibility)
+            container.symbol_table.remove(kernel_sym)
+            routine.name = adj_kernel_sym.name
+        else:
+            routine.name = routine.symbol_table.next_available_name(
+                _create_adjoint_name(routine.name))
 
-    logger.debug("AD kernel will be named '%s'", routine.name)
+            logger.debug("AD kernel will be named '%s'", routine.name)
```

Comment:
Should this be un-indented one level - at the moment it's under the `else` clause?

Reply:
Yes, well noticed.